[Runtime] Extend Graph Runtime To Support Cuda Graph Launch

zhuochenKIDD opened the pull request:
We are currently using graph runtime to run some CTR models on NV-GPU, for our in-house model (around 100 nodes in tvm json graph ) cuGraphLaunch can reduce 5% to 10% percent latency vs the original for-loop cuda kernel launch.

So I wonder if the extension might benefits other workloads, I haven't test other types of models.

comaniac commented:
@zhuochenKIDD is this ready for review? Please modify the description if so; otherwise please mark this PR as a draft first. Thanks.

zhuochenKIDD commented:
@comaniac I've added test case, would you please help review, thanks.

comaniac commented on tests/python/unittest/test_runtime_graph_cugraph.py:
```diff
@@ -0,0 +1,92 @@
+        mlib = tvm.build(s, [A, B], "cuda", name="myadd")
+        ctx = tvm.gpu(0)
+        try:
+            mod = cugraph_runtime.create(graph, mlib, ctx)
```
As I mentioned before, can we use the unified GraphModule interface?

zhuochenKIDD replied:
I added a test test_cuda_graph_runtime in test_runtime_module_based_interface.py to show the use of GraphModule interface, pls check.  We currently use the obsolete way and did't notice the change.

comaniac replied:
No worries. The added interface looks good.

comaniac commented on tests/python/unittest/test_runtime_graph_cugraph.py:
```diff
@@ -0,0 +1,92 @@
+
+
+@tvm.testing.requires_cuda
+def test_graph_simple():
```
You may need to check if CuGraph is enabled and skip this test if not.

zhuochenKIDD replied:
added requires_cudagraph

comaniac commented on CMakeLists.txt:
```diff
@@ -36,6 +36,7 @@ tvm_option(USE_LLVM "Build with LLVM, can be set to specific llvm-config path" O
 tvm_option(USE_STACKVM_RUNTIME "Include stackvm into the runtime" OFF)
 tvm_option(USE_GRAPH_RUNTIME "Build with tiny graph runtime" ON)
 tvm_option(USE_GRAPH_RUNTIME_DEBUG "Build with tiny graph runtime debug mode" OFF)
+tvm_option(USE_GRAPH_RUNTIME_CUGRAPH "Build with tiny graph runtime cuGraph launch mode" OFF)
```
```suggestion
tvm_option(USE_GRAPH_RUNTIME_CUGRAPH "Build with tiny graph runtime with cuGraph for GPUs" OFF)
```

comaniac commented on CMakeLists.txt:
```diff
@@ -321,6 +322,16 @@ if(USE_GRAPH_RUNTIME)
       PROPERTIES COMPILE_DEFINITIONS "TVM_GRAPH_RUNTIME_DEBUG")
   endif(USE_GRAPH_RUNTIME_DEBUG)
+
+  if(USE_CUDA)
+    if(USE_GRAPH_RUNTIME_CUGRAPH)
```
This makes USE_GRAPH_RUNTIME_CUGRAPH silent when CUDA is OFF and may confuse users. We should have
```
if(USE_GRAPH_RUNTIME_CUGRAPH)
  if(NOT USE_CUDA)
    // error out saying please config with USE_CUDA=ON.
```

zhuochenKIDD replied:
I moved this to CUDA.cmake to better check CUDA version > 10, so it might like cudnn/cublas feature, is that ok?

comaniac commented on python/tvm/contrib/cu_graph/cugraph_runtime.py:
```diff
@@ -0,0 +1,62 @@
+    def capture_cuda_graph(self):
+        self._run()  # call cuModuleLoadData before cudaStream API
+
+        print("====== Start Stream Capture ======")
```
Remove unnecessary prints (ditto to the rest).

comaniac commented on python/tvm/contrib/cu_graph/cugraph_runtime.py:
```diff
@@ -0,0 +1,62 @@
+    assert isinstance(graph_json_str, string_types)
+    try:
+        ctx, num_rpc_ctx, device_type_id = graph_runtime.get_device_ctx(libmod, ctx)
+        if num_rpc_ctx == len(ctx):
+            pass
+        else:
+            fcreate = tvm._ffi.get_global_func("tvm.graph_runtime_cugraph.create")
```
```suggestion
        if num_rpc_ctx != len(ctx):
            fcreate = tvm._ffi.get_global_func("tvm.graph_runtime_cugraph.create")
```

comaniac commented on python/tvm/contrib/cu_graph/cugraph_runtime.py:
```diff
@@ -0,0 +1,62 @@
+    except ValueError:
+        raise ValueError(
+            "Please set '(USE_GRAPH_RUNTIME_CUGRAPH ON)' in "
+            "config.cmake and rebuild TVM to enable cu_graph test mode"
```
Why test mode?

zhuochenKIDD replied:
It is because we are currently evaluating CUDA graph API vs kernel launch, and it's keep on going, using TVM is more convenient to do so on new workloads than TF Runtime. And also currently only Kernel-kind cuda node is in captured CUDA graph, in might be more benefits when Memcpy-kind node or using manually created cuda graph, so currently I am not sure current stream-capture way is the optimal way, perhaps need more test

comaniac replied:
I see. We usually call it "experimental". I'll suggest the following:
```
To enable CuGraph (experimental), please set '(USE_GRAPH_RUNTIME_CUGRAPH ON)'
in config.cmake and rebuild TVM
```

comaniac commented on src/runtime/graph/cugraph/graph_runtime_cugraph.cc:
```diff
@@ -0,0 +1,116 @@
+
+class GraphRuntimeCuGraph : public GraphRuntime {
+ public:
+  int StartCapture() {
```
docstring (ditto to the rest).

comaniac commented on src/runtime/graph/cugraph/graph_runtime_cugraph.cc:
```diff
@@ -0,0 +1,116 @@
+
+    CUDA_CALL(cudaStreamBeginCapture(static_cast<cudaStream_t>(capture_stream_),
+                                     cudaStreamCaptureModeGlobal));
+    return 0;
```
If 0 is the only return value, we could simply return void.

comaniac commented on src/runtime/graph/cugraph/graph_runtime_cugraph.cc:
```diff
@@ -0,0 +1,116 @@
+  PackedFunc GetFunction(const std::string& name, const ObjectPtr<Object>& sptr_to_self);
+
+ private:
+  TVMStreamHandle capture_stream_;
```
docstring.

comaniac commented on tests/python/unittest/test_runtime_graph_cugraph.py:
```diff
@@ -0,0 +1,92 @@
+        mod.capture_cuda_graph()
+        a = np.random.uniform(size=(n,)).astype(A.dtype)
+        mod.set_input(x=a)
+        mod.run_cuda_graph()
```
We should think about the user inferface a bit more. For example, can we use the unified `run` API as graph_runtime, and automatically call `capture_cude_graph` before the first run?

zhuochenKIDD replied:
I added  CUDA graph-capture & graph-run control in run(), so for convenient user can just call run(), if user wants manually control, can use capture_cuda_graph + run_cuda_graph

comaniac replied:
Make sense.

comaniac commented on src/runtime/graph/graph_runtime_factory.cc:
```diff
@@ -130,6 +138,31 @@ Module GraphRuntimeFactory::DebugRuntimeCreate(const std::vector<TVMContext>& ct
+Module GraphRuntimeFactory::CudaGraphRuntimeCreate(const std::vector<TVMContext>& ctxs) {
+  const PackedFunc* pf = tvm::runtime::Registry::Get("tvm.graph_runtime_cugraph.create");
+  ICHECK(pf != nullptr) << "Cannot find function tvm.graph_runtime_cugraph.create in registry. "
+                           "Do you enable cuda graph runtime build?";
```
```suggestion
                           "Did you set(USE_GRAPH_RUNTIME_CUGRAPH=ON)?";
```

comaniac reviewed (changes requested):
Just some miner changes but overall is good. Two additional points:
1. I found two terms `cuGraph` and `CUDA graph` are used in this PR. It would be better to just use `cuGraph`.
2. It would be great if you could send a follow-up PR for a tutorial to explain how to use the two interfaces.

zhuochenKIDD commented on cmake/modules/CUDA.cmake:
```diff
@@ -65,6 +65,17 @@ if(USE_CUDA)
+    if(CUDAToolkit_VERSION_MAJOR LESS "10")
+      message(FATAL_ERROR "CUDA Graph requires at least CUDA 10, got=" ${CUDAToolkit_VERSION})
+    endif()
+    message(STATUS "Build with Graph runtime cuGraph support...")
```
Yes I removed all cuGraph or cu_graph name, use CUDA Graph instread

zhuochenKIDD commented:
> Just some miner changes but overall is good. Two additional points:
> 
> 1. I found two terms `cuGraph` and `CUDA graph` are used in this PR. It would be better to just use `cuGraph`.
> 2. It would be great if you could send a follow-up PR for a tutorial to explain how to use the two interfaces.

1. I removed `cuGraph` and changed code to `CUDA graph` because it's NV official terminology and found cuGraph is another lib for graph algorithms
2. I will add more docs when ready, by tutorial do you mean I add a py in tutorials/frontend or a rst in docs/dev?

zhuochenKIDD commented on python/tvm/contrib/nvcc.py:
```diff
@@ -349,6 +349,19 @@ def have_tensorcore(compute_version=None, target=None):
+            return False
+        return True
+    except RuntimeError:
+        warnings.warn("Cannot find cuda path")
```
removed

comaniac commented:
Thanks @zhuochenKIDD

hogepodge commented:
Agree with Tianqi. A how-to guide would be best. You can write it as a Sphinx-Gallery document, under the tvm/tutorials directory. I'm not entirely certain which subdirectory it should go under (you should avoid the get_started directory). Maybe a new directory if it doesn't fit into classifications for the others.
